Add "[" and "]" commands.

a2aaron opened the pull request:
This PR adds "[" and "]" commands to the interpreter.

a2aaron committed:
Add "[" and "]" commands.

"[" Will jump past the matching "]" if the current cell is zero. Otherwise, it does nothing. "]" will jump back to the matching "[" if the current cell is not zero. Otherwise, it does nothing. This enables while loops in BF.

Because flow control means we will be running a BF program for an unknown amount of time, the for loop has been changed to a while loop which exits after reacbing the end of the program.In addition, "program_string" was changed to "program" because it's cleaner looking.

a2aaron committed:
Add extra tests for "[" and "]"

Just some general extra tests. In particular, tests for starting with custom indexes or cell arrays have been added.

a2aaron committed:
Merge branch 'develop' into feature/flow-control

Specificlly, merge the "," and "." additions into this branch. This was done so I don't have to preform conflict resolution on the develop branch.

a2aaron committed:
Add get_brace_matches().

This function takes a BF program and returns a dictionary containing the indexes of each matching brace. "[" is stored as the key and "]" is stored as the value.

a2aaron committed:
Change eval_program() to use a brace dictionary.

eval_program now uses get_brace_matches in "[" and "]" commands.

porglezomp commented on brain_friendly.py:
```diff
@@ -32,31 +32,20 @@ def eval_program(program, index, array, input=None, output=None):
-            else:
-                # Fall out of loop, pop matching "["
-                bracket_stack.pop()
+                for left, right in brace_dict.iteritems():
```
Instead of doing this terrible linear search, store both `left: right` and `right: left` in the dictionary, so you can do `program_index = brace_dict[program_index]` for both directions.

porglezomp replied:
Also, in Python 3, `iteritems` doesn't exist, it's just `items`. Since you're removing this code, that's irrelevant.

a2aaron replied:
Fixed, brace_dict now stores both `left: right` and `right: left`.

a2aaron committed:
Change get_brace_matches to include right brace keys and left brace values.

Before, we were just storing left brace keys to right brace values. Now we store both ways. This allows us to not preform a linear search when trying to jump backwards in a "]" command. Note that this does mean we store twice as many key-value pairs.

Tests have been updated to reflect this new change.

porglezomp commented on test_brain_friendly.py:
```diff
@@ -101,3 +157,28 @@ def test_io_roundtrip():
+
+def test_invalid_braces():
+    # This seems to be the only way to test if multiple expressions raise errors.
+    with pytest.raises(ValueError):
```
I recommend trying something like:

``` python
test_cases = ['[', '[[', ']', ']]', '][', '[[]', '[[[]']
for test_case in test_cases:
    with pytest.raises(ValueError):
        get_brace_matches(test_case)
```

porglezomp commented on test_brain_friendly.py:
```diff
@@ -101,3 +157,28 @@ def test_io_roundtrip():
+
+# Test brace matching
+def test_trival_braces():
+    assert get_brace_matches("") == dict() 
```
You can just do `{}` for an empty dictionary.

porglezomp commented on brain_friendly.py:
```diff
@@ -45,6 +60,29 @@ def decrement_command(value):
+            brace_dict[left_index] = index
+            brace_dict[index] = left_index
+
+    if len(brace_stack) == 0: #sanity check
```
For style, this check should be `if not brace_stack:`, but this would actually probably be better as

``` python
if brace_stack:
    raise ValueError('Malformed BF program (unmatched "["')
return brace_dict
```

porglezomp commented on brain_friendly.py:
```diff
@@ -45,6 +60,29 @@ def decrement_command(value):
+    brace_dict = dict()
+    brace_stack = list()
+
+    for index in range(len(program)):
```
Do this as

``` python
for index, instruction in enumerate(program):
    if instruction == '[':
        brace_stack.append(index)
    # ...
```

porglezomp commented on brain_friendly.py:
```diff
@@ -45,6 +60,29 @@ def decrement_command(value):
+    If the program is malformed, then ValueError is raised.
+    '''
+    brace_dict = dict()
+    brace_stack = list()
```
This should be `brace_stack = []`

porglezomp commented on brain_friendly.py:
```diff
@@ -45,6 +60,29 @@ def decrement_command(value):
+    Returns a dictionary containing the pairs for left and right brace indexes.
+    If the program is malformed, then ValueError is raised.
+    '''
+    brace_dict = dict()
```
This should be `brace_dict = {}`

porglezomp commented on brain_friendly.py:
```diff
@@ -1,13 +1,16 @@
+def eval_program(program, index, array, input=None, output=None):
+    program_index = 0
+    # Stack for "[" and "]" commands.
+    brace_dict = get_brace_matches(program)
```
`brace_dict` isn't the most descriptive name, `brace_associations` might be better? `brace_pairs`?

porglezomp commented on brain_friendly.py:
```diff
@@ -1,13 +1,16 @@
+    program_index = 0
+    # Stack for "[" and "]" commands.
+    brace_dict = get_brace_matches(program)
+    while program_index != len(program):
```
`while program_index < len(program)`, just for safety and clarity.

porglezomp commented on brain_friendly.py:
```diff
@@ -1,13 +1,16 @@
-
-def eval_program(program_string, index, array, input=None, output=None):
-    for command in program_string:
+def eval_program(program, index, array, input=None, output=None):
```
Should we change the name of `array` to `memory` or something?

porglezomp commented on brain_friendly.py:
```diff
@@ -45,6 +60,29 @@ def decrement_command(value):
+        if program[index] == "[":
+            brace_stack.append(index)
+        elif program[index] == "]":
+            if len(brace_stack) == 0: #sanity check
```
`if not brace_stack:`

porglezomp commented on brain_friendly.py:
```diff
@@ -45,6 +60,29 @@ def decrement_command(value):
+            if len(brace_stack) == 0: #sanity check
+                raise ValueError("Malformed BF program (too many \"]\"s)")
+            left_index = brace_stack.pop()
+            #add left AND right brace pairs.
```
Comments should begin with a space and a capital letter

porglezomp commented on test_brain_friendly.py:
```diff
@@ -1,8 +1,11 @@
+import pytest
 import brain_friendly
 from brain_friendly import eval_program
+from brain_friendly import get_brace_matches
```
Change these to `from brain_friendly import eval_program, get_brace_matches`

porglezomp commented on test_brain_friendly.py:
```diff
@@ -1,8 +1,11 @@
+import pytest
```
`import pytest` belongs with `import io`, not with all of our `brainfriendly` imports.
1. Importing "parts of yourself" come first (so the thing you're testing)
2. Importing standard library modules come second
3. Importing external libraries come third
4. Importing local modules (other parts of your library) come fourth

These should probably all be separated by a blank line, although perhaps not 2 and 3.

a2aaron committed:
Add style edits and a few refactors.

Almost all of the changes here are minor style fixes or refactoring varible names. No major program flow alterations.

porglezomp commented:
Looks good, signing off on this!
:+1: :tada: :fireworks: :smile:
